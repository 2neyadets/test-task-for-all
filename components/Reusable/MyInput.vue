<template lang="pug">
  label.field.row.no-wrap.items-start(
    :for="id"
    :class="{ 'field--focused': active, 'field--float': active || value }"
  )
    .field__inner.relative-position.col.self-stretch.column.justify-center
      .field__control.relative-position.row.no-wrap(tabindex="-1")
        .field__control-container.col.relative-position.row.no-wrap
          input.field__native(
            :id="id"
            :ref="'input' + id"
            :value="value"
            @input="$emit('input', $event.target.value)"
            @focus="active = true"
            @blur="active = false"
            type="text"
            tabindex="0"
            :maxlength="maxlength"
          )
          .field__label.no-pointer-events.absolute {{label}}
        template(v-if="clearable && value")
          .q-pl-xs.row.no-wrap.items-center(style="font-size: 24px;")
            button.field__focusable-action.material-icons.notranslate(
              @click="$emit('input', null)"
              @focus="active = true"
              class="material-icons"
              aria-hidden="true"
              role="presentation"
              tabindex="0"
              type="button"
            ) close
      template(v-if="maxlength")
        .field__bottom.row.items-start
          .field__messages.col(v-if="hint") {{hint}}
          .field__counter {{value ? value.length : 0}} / {{maxlength}}
</template>

<script>
import { getId } from '../../utils/helpers'

export default {
  name: 'MyInput',
  components: {},
  props: {
    value: {
      type: [String, Object],
      default: null,
    },
    label: {
      type: String,
      required: true,
    },
    clearable: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    maxlength: {
      type: Number,
      default: 0,
    },
    hint: {
      type: String,
      default: ' ',
    },
  },
  data () {
    return {
      id: getId(),
      active: false,
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {
    if (this.autofocus) this.$refs['input' + this.id].focus()
  },
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .field

    &__native
      line-height 24px
      color #fff
      width 100%
      min-width 0
      outline 0 !important
      font-weight 400
      letter-spacing 0.00937em
      text-decoration inherit
      text-transform inherit
      border none
      border-radius 0
      background none
      padding: 14px 0 2px

    &__inner
      text-align left

    &--focused .field__control:after
      border-color currentColor
      border-width 2px
      transform scale3d(1, 1, 1)

    &__control
      border-radius 0 !important
      height 40px
      padding 0 12px
      color var(--q-color-primary)
      max-width 100%
      outline none

      &:before
        border: 1px solid rgba(255, 255, 255, 0.6)
        transition border-color 0.36s cubic-bezier(0.4, 0, 0.2, 1)

      &:hover:before
        border-color: white

      &:after
        height inherit
        border-radius inherit
        border 2px solid transparent
        transition border-color 0.36s cubic-bezier(0.4, 0, 0.2, 1)

      &:before, &:after
        content ''
        position absolute
        top 0
        right 0
        bottom 0
        left 0
        pointer-events none

    &:not(.field--focused) .field__label, .field__focusable-action
      color rgba(255, 255, 255, 0.7)

    &__label
      font-size 14px
      left 0
      right 0
      top 10px
      color rgba(0, 0, 0, 0.6)
      line-height 20px
      font-weight 400
      letter-spacing 0.01em
      text-decoration inherit
      text-transform inherit
      transform-origin left top
      transition transform 0.36s cubic-bezier(0.4, 0, 0.2, 1), right 0.324s cubic-bezier(0.4, 0, 0.2, 1)

    &--float .field__label
      transform translateY(-30%) scale(0.75)
      right calc(-100% / 3)
    &--focused .field__label
      color currentColor

    &__focusable-action
      opacity 0.6
      cursor pointer
      outline 0 !important
      border 0
      color inherit
      background transparent
      padding 0

    &__bottom
      font-size 12px
      min-height 12px
      line-height 1
      padding 6px 8px
      transform translateY(100%)
      position absolute
      left 0
      right 0
      bottom 0
    &__messages
      line-height 1
    &__counter
      padding-left 8px
      line-height 1
</style>
