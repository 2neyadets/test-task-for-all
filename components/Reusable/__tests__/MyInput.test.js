import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import MyInput from '../MyInput'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('MyInput component', () => {
  let wrapper
  const mockId = '1482363367071'

  beforeEach(() => {
    const label = 'Название'
    wrapper = shallowMount(MyInput, {
      propsData: {
        mockId,
        label,
      },
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('can work with null', done => {
    wrapper.setProps({
      value: null,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('input').element.value).toBe('')
      done()
    })
  })

  it('can work with undefined', done => {
    wrapper.setProps({
      value: undefined,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('input').element.value).toBe('')
      done()
    })
  })

  it('can work with strings', done => {
    const value = 'value123123'
    wrapper.setProps({
      value,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('input').element.value).toBe(value)
      done()
    })
  })

  it('is clearable', done => {
    const value = 'someValue'
    wrapper.setProps({
      value,
      clearable: true,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('button').exists()).toBe(true)
      done()
    })
  })

  it('input emit "input" event', () => {
    const value = '111!!!'
    wrapper.find('input').setValue(value)
    expect.assertions(2)
    expect(wrapper.find('input').element.value).toBe(value)
    expect(wrapper.emitted().input[0]).toEqual([value])
  })

  it('clear btn emit "input" event by clicking on it', done => {
    const value = 'someValue'
    wrapper.setProps({
      value,
      clearable: true,
    })
    localVue.nextTick(() => {
      wrapper.find('button').trigger('click')
      expect.assertions(1)
      expect(wrapper.emitted().input[0]).toEqual([null])
      done()
    })
  })

  it('can show max length', done => {
    const value = 'someValue',
      maxlength = 10
    wrapper.setProps({
      value,
      maxlength,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('.field__counter').html()).toContain(`${value.length} / ${maxlength}`)
      done()
    })
  })

  it('can show hint', done => {
    const hint = 'Подсказка!'
    wrapper.setProps({
      hint,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.find('.field__messages').html()).toContain(hint)
      done()
    })
  })

  it('can autofocus', () => {
    const localWrapper = shallowMount(MyInput, {
      propsData: {
        label: 'Название',
        autofocus: true,
      },
    })
    expect.assertions(1)
    expect(localWrapper.find('input').element).toBe(document.activeElement)
  })

  it('can be active and not', done => {
    expect.assertions(6)
    expect(wrapper.classes('field--focused')).toBe(false)
    expect(wrapper.classes('field--float')).toBe(false)
    wrapper.find('input').trigger('focus')
    localVue.nextTick(() => {
      expect(wrapper.classes('field--focused')).toBe(true)
      expect(wrapper.classes('field--float')).toBe(true)
      wrapper.find('input').trigger('blur')
      localVue.nextTick(() => {
        expect(wrapper.classes('field--focused')).toBe(false)
        expect(wrapper.classes('field--float')).toBe(false)
        done()
      })
    })
  })

  it('id on checkbox and label', () => {
    expect.assertions(2)
    expect(wrapper.find('label').attributes('for')).toBe(mockId)
    expect(wrapper.find('input').attributes('id')).toBe(mockId)
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
