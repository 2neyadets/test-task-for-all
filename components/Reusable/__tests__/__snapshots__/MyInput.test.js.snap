// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyInput component match snapshot 1`] = `
<label
  class="field row no-wrap items-start"
  for="1482363367071"
>
  <div
    class="field__inner relative-position col self-stretch column justify-center"
  >
    <div
      class="field__control relative-position row no-wrap"
      tabindex="-1"
    >
      <div
        class="field__control-container col relative-position row no-wrap"
      >
        <input
          class="field__native"
          id="1482363367071"
          maxlength="0"
          tabindex="0"
          type="text"
        />
        <div
          class="field__label no-pointer-events absolute"
        >
          Название
        </div>
      </div>
      <!---->
    </div>
    <!---->
  </div>
</label>
`;
